### DISCUSSION TITLE
Environmentalized deployments and config-file based approach

### DISCUSSION THREAD

**Discussion by WebVoid** [directus/directus#17495] [Feature Requests]
# Introduction and motivation

Hi fellow developers and open-source lovers

## As a disclaimer

I did not have the time to go very in-depth into the architecture of the solution, and maybe am missing some things. I write this message very respectfuly and expect from the community to answer likewise.

## Context

I work at a company where we are on the verge of starting a lot of projects for internal usage and my boss asked me to look into Directus. Although I love the solution, I have some concerns concerning the industrialization capability of directus-based applications, namely being able to actually develop something, push it on a repo, merge multiple developers work and then deploy the changes to successive environments (we work with Docker, k8s and AWS).

Maybe I'm an idiot and did not find the right way, but **if** I'm not wrong, the only actually exportable and importable things today are **schemas**.

The problem is that considering that you have :

- Your **app data schema** and **app data**
- The **Directus data schema**
- The **Directus project data (config and such)**

For the first one, no problem, you can export from an environment and import on another one. Bonus: you can actually version it, and it is kind of merge friendly (yeay).

For the second one, no problem, it is to be managed by the Directus instance itself when migrating database schemas.

For the latest, the problem is that it includes two things, the **project settings**, like name, url, etc... that **can** be considered environment-dependant (altho you may want to environmentalize them), and then things like flows, operations, roles and permissions, things that can be considered **actual application logic**, and that we therefore would want to be able to have multiple developpers working on, and then be able to version, merge and deploy across all our env successively.

By my understanding (once again, maybe I'm an idiot), there is no way right now to export, merge, version and import all those things that can be considered part of the application logic, except by doing database export versionning and import which is neither very dev nor devops friendly.

# Pet vs Cattle

A quick reminder for those who would not understand the problem I'm trying to explain.

When you have an application that you deploy, configure directly on the end server, and then you backup, are able to restore, but not to actually environmentalize the config and actual app logic, allowing you to easily create new environments for this project, you are keeping a pet on a leash. You love it, you take care of it, but the day it dies (or turns against you), you're in big trouble.

This is the case for instance for Wordpress, Prestashop and all those all-in-one-cms-with-a-big-wonderful-admin-panel-based applications.

On actual projects, you have many people developping on the same application, commiting, merging, deploying new **app logic** to a dev or staging environment for validation, and then when you get the approval, you deploy to prod. You may even want to have feature branch deployments, allowing you to spin up environments like pop corn in a pan for demo or e2e testing purposes.

The problem is that if this app logic is part of the application and not configurable and describable with **pure code**, you can't spin up pre-configured reproductible environments that easily.

# I did try

I went through a lot a discussions on the Directus Github repository and have seen a lot of very interesting people talk about this exact problem, trying to come up with solutions I respectfuly do not think cover the problem in it's entirety.

Sometimes it is about using environment variables, sometimes about config files for two parameters or so, but that never seemed to fully cover the problem, just to answer the immediate need of a precise scenario.

# The way I see it

> Config files. For everything. Really.

A config-file-first approach would allow to have scenarios where :

- A dev starts working on the backend, with a Dockerized or npx-started directus, in a working directory
- He does the basic app config, logic, schemas, extensions, etc... (or starts with an existing database schema)
- As he does things, everything that can be considered app logic is saved into the working directory as config files (json or more likely yaml), alongside other files like extensions, panels, etc... (that does not exclude saving parameters the database)
- When he commits, everything is there, every last bit of information another instance of Directus would need to be spinned up with the same logic
- In the next days, other developpers clone the directory, spin up their local Directus, which reads the config files and start up already pre-configured
- When they do changes, the changes are written to the config files, with the extension code or panels they could do
- When they commit, **this allows for atomic changes to be commited separately**, and for **git merge requests to be easily processable** (and also feature environments, integration testing, etc... to be possible)
- When the changes are accepted, they are deployed to a dev (or staging, or anything) environment, the Directus application starts/restarts and imports the new config, then starts in a semi readonly mode, allowing only for application data to be changed, but nothing in the settings except for environment-specific data like users
- Now you can have your config, extensions, plugins, panels and everything going through all your environments and be reproductible and testable at every step

Notes:

- This would not prevent everything that is doable today to work, plus it could be an option, enabled by an environment variable or so.
- By having multiple env variables we could choose not to environmentalize certain app parameters, the minimum for me being roles and permissions, flow and operations, dashboards and panels, presets and finally webhooks.
- This approach would also allow in some cases **configuration files templating**, allowing for a CI/CD pipeline to replace certain values depending on the environment you are deploying onto, or for testing purposes.

**Comment by rijkvanzanten** [comment]
That third point is a point of much debate! 

> The Directus project data (config and such)

Cause often, not _all_ project data is considered part of the versioned / deployable setup, and the "and such" is highly dependent on project. 

Just a couple spitfire questions to fully understand some more of the edge case angles here:

- Would you effectively "overwrite" an entire table from the versioned file? 
- Would you envision that the API effectively serves from the file directly, skipping the database?
- Would you still want to be able to alter the content/data from Directus, and then make a new snapshot?
- Would this file potentially have a partial "dump"? Eg when thinking about Directus Permissions, would you envision being able to "snapshot" the permissions for a particular role, but not all?
- Would this only be for Directus configuration, or would you image custom tables to be part of this system as well? (Fe a `config` table that you use for your own apps)

You already touched on some notes that overlap with some of those questions, but I'd love to discuss this in more detail, as it's something I've been struggling with as well, but have had trouble trying to find a good solution for that'll actually work for the various use cases this touches ✨

**Reply by WebVoid** [reply]
For the first point, when I say Directus project data, I'm talking really about the data Directus stores in its db that is not at all related with the actual application data (yours), I'm talking about the data present in the DB tables `dashboard`, `operations`, `panels`, `permissions`, `presets`, `roles`, `webhooks` and eventually `settings`

For your questions, in order:
- Yes the entire table would be truncated and then populated with the config from the files, this is the only way to have actually reproductible environments you can spin from void to production just from a git repository
- It could be possible that the database disapears, but still having the data in the database allows progressive adoption of config files whilst still using some parts of the config from the database
- Yes, you would be able to alter the config of your Directus instance but only in a local/dev environment, the "on-actual-server" environments would be kind of read only to prevent unwanted modifications to occur and most importantly not being reported to the config stored on maybe git or a single source of trust
- I would not recommend snapshotting a particular subset of a data for reproduction's sake, I know that we might sometimes want to have some particular behavior depending on the environment of the application, but I would still not recommend doing it directly on the instance, which is why I talked in my first message about config file templating, which would allow this use case to occur (if you're into DevOps, it could be just like either kustomize for Kubernetes, allowing you to override subsets of your config with another config file or Helm, with mustached parameters from a file, conditionals and everything)
- I'm not talking here about things that touch your actual app config, only the Directus things, but the resulting work could be exposed as an API (JS, not http) for anyone to be able to write an extension to read a config file from the fs and wipe a table and import the content of the file with a one-liner (like Directus::overrideConfigToTable(<my_conf_file>, <my_table>))

I'd love to discuss more about the subject if you have any more question =)

**Reply by rijkvanzanten** [reply]
Awesome! The tricky bit to figure out is what parts are moved and what parts aren't. Not everybody considers all of those tables to be part of the things that should be versioned (but that's where additional config comes in). The main question remaining to me is whether or not partial 'snapshotting'[^1] should be available in some capacity. The main reason I mention that is that it's not uncommon to have dev-only additional things in the environment that you don't want to move over, for example you could have an additional dev-only admin role just to test things with, that you want to exclude from the versioned updates 🤔 

[^1]: When I say "snapshot", I mean effectively an "snapshot" of the current state of the data, but in a readable versioning-friendly format

**Reply by WebVoid** [reply]
It could be possible, but would be harder to implement for one would have to choose which data to import and export. I think this is a problem that can be resolved by hard splitting configs into something like one file per DB line or so (for this kind of data) and then tricking the CI/CD pipelines to add/remove those files at deployment time

**Reply by rijkvanzanten** [reply]
True! Although exporting could already rely on the existing filtering capabilities, which solves for that part. If import is a more "everything or nothing", I think we can strike a nice balance 🙂

**Reply by WebVoid** [reply]
My point is I wouldn't want developers to actually have some action to do in order to export their config, as manual actions tend to be a source of errors, which is why partial export is for me hard to setup

**Comment by rogier-mars-mck** [comment]
Hi all!

Thanks @WebVoid for the elaborate description of this request. I've been in contact with Directus support regarding the same ask and eventually have been spiking out a solution that works for me. I'll try to share this approach as it might help you move forward.

Upon starting my container, the following steps are performed:

1) Directus bootstrap
    - Installs the database
    - Runs database migrations (internal and extensions)
    - Creates first admin user and role
2) Administrator uuid fix: this custom script ensures the uuid of the initial admin role is consistent across environments. Notice that this cannot happen as part of a migration because the first admin user and role creation happens after running the migrations in the bootstrap step.
3) Applying snapshots
    - Files and Folders: only "local" storage files that are used for the admin UI
    - Data Model: collections and custom fields
    - Roles & Permissions
    - Presets
    - Project Settings
4) Directus start: running the application

Parts involved:
- extensions/migrations:
    - uuid extension
    - adding uuid "external_id" field to tables that use incremental id: permissions, presets, project-settings
    -  creating an admin folder
- custom scripts:
    - administrator-uuid-fix
    - snapshot import and exports:
        - files-folders
        - presets
        - project-settings
        - roles-permissions
 
Notice that the project-settings also contain assets: the images used for the logo and foreground. I've came up with a solution where these images are part of the upload folder (check into git) and make use of the "local" storage adapter. Regular collection assets could make use of a different storage adapter such as aws-s3.

Hope this helps!

**Reply by WebVoid** [reply]
When you talk about snapshots, are those database snapshots ? If so, this is a point I discussed in my notebook but I find it really not dev nor devops friendly because they are not **really** version-control friendly.
The thing is, it works, but you can't have multiple devs working on the same Directus instance, developing backend logic for different parts of the application and then merging everything on a source-control directory like Git or SVN to have an actual single source of trust...

The approach is interesting tho

**Reply by rogier-mars-mck** [reply]
I'm not talking about database snapshots, sorry if that was unclear.

I've went with the approach Directus uses for data-model "snapshots": https://docs.directus.io/self-hosted/cli.html#migrate-schema-to-a-different-environment

Import from and export to a yaml file. These files are checked into Github.

**Reply by WebVoid** [reply]
Yes that is more clear, this is the way to go I think, as I explain this is the only way for me to have some actual app configuration versioning and reproduction method

**Reply by rijkvanzanten** [reply]
> Notice that the project-settings also contain assets: the images used for the logo and foreground.

This is a great point, we'll have to consider assets as part of this bigger conversation as well 👍🏻

**Comment by dmatthams** [comment]
Hey! I've just started setting up a development workflow for the first time with Directus and have come across the same issue here. I'm unsure of the best way to handle any changes to roles, flows etc – and I don't really want to be doing this in production.

I also use Craft CMS and they've got a really nice solution (but obviously what works for them might not work here). *All* settings are stored in 'project config' yaml files, which are automatically generated as the database schema changes. Then on each deploy they are applied to the production database. On the production environment **there is no access to the settings whatsoever**, just the content. Project logo, background etc. would just be checked in to git and environment variables are used for any environment-specific settings.

So the development workflow is simply: pull remote database ➡ update project settings locally ➡ push config files to sync changes to production.

**Reply by rijkvanzanten** [reply]
That sounds very close to what's discussed above as well. That being said, I'm also wondering if we can somehow combine that with a "code-first" approach like the suggestion of @exsurgo in https://github.com/directus/directus/discussions/13041#discussioncomment-5095848 🤔

**Reply by dmatthams** [reply]
Purely from my perspective I want to use Directus *because* of the intuitive UI, I can't see a situation where I'd want to build the schema in code. Yaml/Json files (split down into smaller chunks) can work in a git workflow, plus they can be used 'code-first' if you wanted to.

**Reply by rijkvanzanten** [reply]
I don't disagree! Just interesting to keep "the other side" of that discussion in mind as well. I can also imagine we'd come up with a separate tool that allows you to write it as "code", which then generates to the same yaml syntax for ease of use.

**Reply by WebVoid** [reply]
This is very interesting indeed and I believe both points of view should be considered, code-first would be more interesting for actual developers, to just rely on Directus for the actual backend engine. Now I also agree with @dmatthams about the fact that Directus's wonderful interface is a go-go for us also, and that it is why I was suggesting a config-first approach, also because yaml files are easily "buildable" in a CD chain, therefore allowing the deployment of pre-configured instances of Directus (block-based approach with modular instances, startable with everything needed for a particular use-case, lego-like, could be an interesting idea for the cloud version also ;) @rijkvanzanten )

**Comment by WebVoid** [comment]
@rijkvanzanten I was thinking that maybe as a first step this could be implemented as cli (or more likely `@directus/cli`) commands like `directus config snapshot ./config` and `directus config apply ./config` which would respectively export/import the configs to/from yaml files. Options could be used to define the categories of settings the user wishes to dump/restore (the list still remains to be established and refined).

This could allow a progressive adoption and ensure not to break anything in the core right now. It would also probably be more secure and fail-safe for it could check for bizarre behaviors and ask the user, whilst still being integrable with automation flows with a `yes` flag.

Also would it be more pertinent to implement this in the client or server cli ? Or maybe both ?
I think it would be more handy in the client cli for the goal is more to use it in a CI pipeline or a desktop rather than directly on the machine that executes it (in cloud setup) but it would need to go through API manipulation rather than direct database manipulation.

What do you think ? I would be interested in proposing an implementation soon if the idea was triaged, for this is something we need before we can move forward with the solution.


Another idea would be to implement this as an open-source module that could be installed in a Directus instance by the user, giving him API endpoints to import/export configs, but not touching at all to the core codebase whilst crashtesting the idea.

**Reply by rijkvanzanten** [reply]
Ideally, we could start with a client-side implementation that relies on the existing rest/graphqhl APIs to do the actual work. That way, this would work for any installation in any hosting environment. (It's not always as easy to use a local CLI if you're in a containerized platform-as-a-service environment)

**Reply by rijkvanzanten** [reply]
> Another idea would be to implement this as an open-source module that could be installed in a Directus instance by the user, giving him API endpoints to import/export configs, but not touching at all to the core codebase whilst crashtesting the idea.

I'd prefer this at the moment! There's so many edge/use cases involved that it's going to be really difficult to implement an initial version that'll be good enough for most. I'd rather take some time trying out approaches outside of core before landing on a single solution that seems to work for the majority which can then be implemented in core 🙂

**Reply by WebVoid** [reply]
> Ideally, we could start with a client-side implementation that relies on the existing rest/graphqhl APIs to do the actual work. That way, this would work for any installation in any hosting environment. (It's not always as easy to use a local CLI if you're in a containerized platform-as-a-service environment)

I agree, which is why I was suggesting implementing it on the client cli

> > Another idea would be to implement this as an open-source module that could be installed in a Directus instance by the user, giving him API endpoints to import/export configs, but not touching at all to the core codebase whilst crashtesting the idea.
> 
> I'd prefer this at the moment! There's so many edge/use cases involved that it's going to be really difficult to implement an initial version that'll be good enough for most. I'd rather take some time trying out approaches outside of core before landing on a single solution that seems to work for the majority which can then be implemented in core 🙂

Also agreed, so in the end do you prefer going for a client cli implementation or a module ? This could also be another cli based on the main one, allowing for easy integration if things go well

**Reply by rijkvanzanten** [reply]
> Also agreed, so in the end do you prefer going for a client cli implementation or a module ? This could also be another cli based on the main one, allowing for easy integration if things go well

Gut-feeling says it'd be easiest as a standalone cli-based tool. @codingbutter just started something that could be a great kicking-off point: https://github.com/CodingButter-LLC/directus-migrator

**Reply by WebVoid** [reply]
After a lot of intense thinking, I will start an implementation in the client CLI by forking the @directus/cli package, I have a feeling that if it goes well it will be easier to integrate afterwards.

The code from @CodingButter is interesting but it is an actual config migrator, it doesn't allow for a dump and restore style management, which is a go-to for pre-configured instances, developers local config sharing and config assertion testing (talking about real-world needs here).

Also I fear that if 10 different CLI tools are available for different use-cases it will be a pain for end-users to find what they want and they may end up thinking that their use-case is not covered and pass their way.

I will post here any update about this subject, also I joined the Discord server as `Wolf [VLF]#4923` if we need to reach out.

**Reply by rijkvanzanten** [reply]
> Also I fear that if 10 different CLI tools are available for different use-cases it will be a pain for end-users to find what they want and they may end up thinking that their use-case is not covered and pass their way.

Oh for sure! I see this more as an way to experiment and do more agile R&D while finding out the "final" solution that can then be added to core as the standardized way 🙂 There have been a lot of questions / concerns around how this should work exactly (including how to select what data is treated 'config' as part of these exports) so this approach should allow us to figure some of those things out before adding it to core :)

**Reply by WebVoid** [reply]
I agree, I will implement it in a fork of the main client cli then

**Comment by rijkvanzanten** [comment]
Lets combine this discussion with https://github.com/directus/directus/discussions/13041! Even though this discussion has some more in-depth kick-off points, I want to make sure we don't lose focus on the main feature (and spread the engagement) between the two that are effectively aiming to achieve the same goal 💪🏻
